independent-projects/bootstrap/runner/src/main/java/io/quarkus/bootstrap/runner/JarFileReference.java

```diff
@@ -52,33 +82,52 @@ private boolean acquire() {
      */
     private boolean release(JarResource jarResource) {
         while (true) {
-            int count = referenceCounter.get();
-            if (count <= 0) {
-                throw new IllegalStateException(
-                        "The reference counter cannot be negative, found: " + (referenceCounter.get() - 1));
+            final int count = referenceCounter.get();
+            // Both 1 and 0 are invalid states, because:
+            // - count = 1 means that we're trying to release a jarFile not yet marked for closing
+            // - count = 0 means that the jarFile has been already closed
+            if (count == 1 || count == 0) {
+                throw new IllegalStateException("Duplicate release? The reference counter cannot be " + count);
             }
-            if (referenceCounter.compareAndSet(count, count - 1)) {
-                if (count == 1) {
-                    jarResource.jarFileReference.set(null);
-                    try {
-                        jarFile.close();
-                    } catch (IOException e) {
-                        // ignore
-                    }
+            if (referenceCounter.compareAndSet(count, addCount(count, -1))) {
+                if (count == -1) {
+                    silentCloseJarResources(jarResource);
                     return true;
                 }
                 return false;
             }
         }
     }
 
+    private void silentCloseJarResources(JarResource jarResource) {
+        // we need to make sure we're not deleting others state
+        jarResource.jarFileReference.compareAndSet(completedFuture, null);
+        try {
+            jarFile.close();
+        } catch (IOException e) {
+            // ignore
+        }
+    }
+
     /**
      * Ask to close this reference.
      * If there are no readers currently accessing the jarFile also close it, otherwise defer the closing when the last reader
      * will leave.
      */
-    void close(JarResource jarResource) {
-        release(jarResource);
+    void markForClosing(JarResource jarResource) {
```

Comment:
This method doesn't really close the resource, but mark it as ready to be closed, now its method better reflect this behaviour.